server/app/query/page.tsx

```diff
@@ -1,11 +1,189 @@
+"use client";
+
+import { useEffect, useState } from "react";
+import Link from "next/link";
+
+import { StatusPill, RunTimePill } from "@/app/query/view/[id]/components";
+import {
+  StatusEvent,
+  RemoteServer,
+  RemoteServerNames,
+  IPARemoteServers, //hack until the queryId is stored in a DB
+  StatusEventByRemoteServer,
+  initialStatusEventByRemoteServer,
+} from "@/app/query/servers";
+import { getQueryByUUID, Query } from "@/data/query";
+
+type QueryData = {
+  statusEvent: StatusEventByRemoteServer;
+  query: Query;
+};
+type DataByQuery = {
+  [queryID: string]: QueryData;
+};
+
 export default function Page() {
+  const [queryIDs, setQueryIDs] = useState<string[]>([]);
+  const [dataByQuery, setDataByQuery] = useState<DataByQuery>({});
+
+  const updateData = (
+    query: Query,
+    remoteServer: RemoteServer,
+    statusEvent: StatusEvent,
+  ) => {
+    setDataByQuery((prev) => {
+      let _prev = prev;
+      if (!Object.hasOwn(prev, query.uuid)) {
+        // if queryID not in dataByQuery yet,
+        // add initial status before updating value
+        _prev = {
+          ..._prev,
+          [query.uuid]: {
+            statusEvent: initialStatusEventByRemoteServer,
+            query: query,
+          },
+        };
+      }
+
+      return {
+        ..._prev,
+        [query.uuid]: {
+          ..._prev[query.uuid],
+          statusEvent: {
+            ..._prev[query.uuid].statusEvent,
+            [remoteServer.remoteServerName]: statusEvent,
+          },
+        },
+      };
+    });
+  };
+
+  useEffect(() => {
+    // poll runningQueries every second
+    (async () => {
+      const interval = setInterval(async () => {
+        const _queryIDs: string[] =
+          await IPARemoteServers[RemoteServerNames.Helper1].runningQueries();
+
+        setQueryIDs(_queryIDs);
+      }, 1000); // 1000 milliseconds = 1 second
+      return () => clearInterval(interval);
+    })();
+  }, []);
+
+  useEffect(() => {
+    (async () => {
+      setDataByQuery((prev) => {
+        return Object.fromEntries(
+          Object.keys(prev)
+            .filter((queryID) => queryIDs.includes(queryID))
+            .map((queryID) => [queryID, prev[queryID]]),
+        );
+      });
+
+      for (const queryID of queryIDs) {
+        const query: Query = await getQueryByUUID(queryID);
+
+        for (const remoteServer of Object.values(IPARemoteServers)) {
+          const statusEvent: StatusEvent =
```

Comment:
is there a way in javascript to fan out these queries instead of traversing all servers sequentially?

Reply:
doesn't the await make that effectively what happens, since this is async?

Reply:
I could be wrong because I don't know the details of async implementation in JS, my understanding is that code does not progress past the `await` point until this request resolves. Calling this in a loop should make 3 sequential requests to 3 servers. 

There needs to be a way to collect these tasks and await them all at once

Reply:
good callout. it looks like `Promise.all` is the right way to do that.